sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PropagateEmptyRelation.scala

```diff
@@ -111,7 +111,8 @@ abstract class PropagateEmptyRelationBase extends Rule[LogicalPlan] with CastSup
           // Except is handled as LeftAnti by `ReplaceExceptWithAntiJoin` rule.
           case LeftOuter | LeftSemi | LeftAnti if isLeftEmpty => empty(p)
           case LeftSemi if isRightEmpty | isFalseCondition => empty(p)
-          case LeftAnti if isRightEmpty | isFalseCondition => p.left
+          case LeftAnti if isRightEmpty && canExecuteWithoutJoin(p.left) => p.left
+          case LeftAnti if isFalseCondition && canExecuteWithoutJoin(p.left) => p.left
```

Comment:
It's simply `(isRightEmpty | isFalseCondition) && canExecuteWithoutJoin...`?

Reply:
updated, thank you.